From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rachapalli Bandi Jagadeesh <jagadeesh.rachapalli.bandi@intel.com>
Date: Thu, 23 Apr 2026 22:58:27 +0530
Subject: Fix class->devnode prototype mismatch

Struct class::devnode as taking const struct device * from 6.2.0 KV.
while above kernels use struct device *.

Add an autotools m4 compile probe to detect the expected devnode prototype and
define BPM_CLASS_DEVNODE_CONST_NOT_PRESENT when const is not used.
Switch vfio_devnode() to use this macro to select the correct function signature
and keep builds across kernel variants without error.

Error:
--------------------------------------------------------------------------------
error: assignment to ‘char * (*)(struct device *, umode_t *)’ {aka ‘char * (*)
(struct device *, short unsigned int *)’} from incompatible pointer type ‘char *
(*)(const struct de  vice *, umode_t *)’ {aka ‘char * (*)
(const struct device *, short unsigned int *)’}
------------------------------------------------------------------------------

Signed-off-by: Rachapalli Bandi Jagadeesh <jagadeesh.rachapalli.bandi@intel.com>
---
 drivers/vfio/group.c     | 8 ++++++++
 drivers/vfio/vfio_main.c | 4 ++++
 2 files changed, 12 insertions(+)

diff --git a/drivers/vfio/group.c b/drivers/vfio/group.c
index b2299e5..4f1d335 100644
--- a/drivers/vfio/group.c
+++ b/drivers/vfio/group.c
@@ -15,7 +15,11 @@
 #include <linux/anon_inodes.h>
 #include "vfio.h"
 
+#ifdef BPM_CLASS_DEVNODE_CONST_NOT_PRESENT
+static char *vfio_devnode(struct device *dev, umode_t *mode);
+#else
 static char *vfio_devnode(const struct device *, umode_t *);
+#endif
 static const struct class vfio_class = {
 	.name	= "vfio",
 	.devnode = vfio_devnode
@@ -885,7 +889,11 @@ bool vfio_file_has_dev(struct file *file, struct vfio_device *device)
 }
 EXPORT_SYMBOL_GPL(vfio_file_has_dev);
 
+#ifdef BPM_CLASS_DEVNODE_CONST_NOT_PRESENT
+static char *vfio_devnode(struct device *dev, umode_t *mode)
+#else
 static char *vfio_devnode(const struct device *dev, umode_t *mode)
+#endif
 {
 	return kasprintf(GFP_KERNEL, "vfio/%s", dev_name(dev));
 }
diff --git a/drivers/vfio/vfio_main.c b/drivers/vfio/vfio_main.c
index 6222376..4d42eaf 100644
--- a/drivers/vfio/vfio_main.c
+++ b/drivers/vfio/vfio_main.c
@@ -63,7 +63,11 @@ MODULE_PARM_DESC(enable_unsafe_noiommu_mode, "Enable UNSAFE, no-IOMMU mode.  Thi
 
 static DEFINE_XARRAY(vfio_device_set_xa);
 
+#ifdef BPM_CLASS_DEVNODE_CONST_NOT_PRESENT
+static char *vfio_device_devnode(struct device *dev, umode_t *mode)
+#else
 static char *vfio_device_devnode(const struct device *dev, umode_t *mode)
+#endif
 {
 	return kasprintf(GFP_KERNEL, "vfio/devices/%s", dev_name(dev));
 }
-- 
2.43.0

